In [8]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
# from sklearn.linear_model import LinearRegression

In [9]:
data = {'so_khach': [61, 63, 67, 69, 70, 74, 76, 81, 86, 91, 95, 97],
        'chi_phi': [4.280, 4.080, 4.420, 4.170, 4.480, 4.300, 
                     4.820, 4.700, 5.110, 5.130, 5.640, 5.560]
       }
table = pd.DataFrame(data)
print(table)

    so_khach  chi_phi
0         61     4.28
1         63     4.08
2         67     4.42
3         69     4.17
4         70     4.48
5         74     4.30
6         76     4.82
7         81     4.70
8         86     5.11
9         91     5.13
10        95     5.64
11        97     5.56


In [10]:
so_khach = sm.add_constant(table.so_khach) #Tạo 1 bảng mới gồm 1 cột hằng số và cột so_khach
# statsmodels.OLS(para1, para2).fit()
# para1: biến phụ thuộc
# para2: biến độc lập
model = sm.OLS(table.chi_phi, so_khach).fit() # OLS: Phương pháp bình phương bé nhất
model.summary()

C:\Users\Admin\AppData\Roaming\Python\Python312\site-packages\scipy\stats\_axis_nan_policy.py:418: UserWarning: `kurtosistest` p-value may be inaccurate with fewer than 20 observations; only n=12 observations were given.
  return hypotest_fun_in(*args, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                chi_phi   R-squared:                       0.899
Model:                            OLS   Adj. R-squared:                  0.889
Method:                 Least Squares   F-statistic:                     89.09
Date:                Fri, 04 Oct 2024   Prob (F-statistic):           2.69e-06
Time:                        09:20:17   Log-Likelihood:                 4.8312
No. Observations:                  12   AIC:                            -5.662
Df Residuals:                      10   BIC:                            -4.693
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5698      0.338      4.643      0.001       0.816       2.323
so_khach       0.0407      0.004      9.439      0.000       0.031       0.050
==============================================================================
Omnibus:                        1.556   Durbin-Watson:                   2.962
Prob(Omnibus):                  0.459   Jarque-Bera (JB):                0.872
Skew:                          -0.253   Prob(JB):                        0.647
Kurtosis:                       1.780   Cond. No.                         518.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [11]:
# cột coef có 2 giá trị lần lượt là B0 và B1
# cột P > |t|: p-value

#### Kết quả

Từ những kết quả trên ta có:

**- Mô hình ước lượng (mô hình hồi quy mẫu) của chi phí theo số khách:**
 $$\hat{Y} = 1.5698 + 0.0407*X$$

 **$\beta_0 = 1.5698$** - *Ý nghĩa: Chi phí trung bình khi số khách = 0 khoảng 1.5698*
  
 **$\beta_1 = 0.0407$** - *Ý nghĩa: Khi số hành khách tăng thêm 1 đơn vị thì trung bình chi phí bỏ thêm tăng khoảng 0.0407*

**- Hệ số xác định của mô hình là $R^2 = R-squared = 0.899$**

**Ý nghĩa của hệ số $R^2$**

- $R^2 = 0.899 > 0.8$ thì mô hình phù hợp cao so với tập só liệu

- 89.9% sự biến động của chi phí (biến phụ thuộc) được giải thích bởi số khách (từ mô hình)

---

##### Khoảng tin cậy 95% cho hệ số góc của mô hình:

so_khach: [0.025, 0.975] = [0.031, 0.050]

---

##### Kiểm định xem số khách ảnh hưởng có ý nghĩa đến chi phí của chuyến bay không tại $\alpha = 0.05$
- Bài toán kiểm định sự có ý nghĩa của mô hình:
$H_0: \beta_1 = 0$ và $H_1: \beta_1 \ne 0$

có p-giá trị là: P>|t| < 0.001 (vì kết quả cho 0.000 đến 3 chữ số sau dấu phảy vẫn là 0 nên nhỏ hơn 0.001 ). Từ đó p-giá trị < 0.05 nên bác bỏ H0 hay mô hình hồi quy có ý nghĩa. Tức là số khách ảnh hưởng có ý nghĩa đến chi phí của chuyến bay

In [12]:
# Dự báo chi phí khi số khách bằng 75
model.predict([1, 75])
# Khi số khách là 75 thì chi phí trung bình của chuyến bay xấp xỉ 4.6224 đơn vị

array([4.62241267])

In [13]:
# Khoảng tin cậy 95% cho hệ số góc
model.conf_int(0.05)

0         1
const     0.816497  2.323089
so_khach  0.031094  0.050310